File: rabin-miller.js
/*
 * Function for modular exponentiation (ensures against integer overflow),
 * computes (base^expo) mod n
 * @param base : used as base
 * @param expo : used as expo
 * @param n : used to mod
 * @return result : result = ((base^expo) mod n)
 */
function modularExponentiation(base, expo, n) {
    let val = 1;
    while (expo > 0) {
        // the case that expo is odd, make sure to multiply by one base
        if (expo % 2 === 1) {
            val = (val * base) % n;
        }
        // repeated squaring
        base = (base * base) % n;
        expo = Math.floor(expo / 2);
    }
    return val;
}

 /*
  * Function implements Rabin Miller Primality Test
  * @param n : number to test for primality
  * @param u : denotes u in the equation (n-1 = (2^t)u)
  * @return bool : if prime True else False
  */
 function RabinMillerTest(n, u) {
    // pick random integer 2 <= a < n
    let a = Math.floor(Math.random() * (n-2)) + 2;
    let expo = u;
    let val;

    while (expo !== n-1) {
        val = modularExponentiation(a, expo, n);
        if (val !== 1 && val !== n-1) {
            console.log(a, u, expo);
            return false;
        }
        expo *= 2;
    }
    val = modularExponentiation(a, expo, n);
    if (val !== 1) {
        console.log(a, u, expo);
        return false;
    }
    return true;
 }

/*
 * Function checks whether a # is prime using Rabin Miller Primality Test
 * @param n : number to test for primality
 * @param numTrials : # of times to run probabilistic Rabin Miller test
 * @return bool : if prime True else False
 */
function isPrime(n, numTrials) {
    if (n < 2) return false;
    else if (n !== 2 && n % 2 === 0) return false;

    // compute u, s.t. n-1 = (2^t)u
    let u = n - 1;
    while (u % 2 === 0) {
        u = Math.floor(u / 2);
    }

    // run trials
    for (let i = 0; i < numTrials; i++) {
        if (!RabinMillerTest(n, u)) {
            return false;
        }
    }
    return true;
}

let n = 294409;
let numTrials = 100;
isPrime(n, numTrials);

File: rabin-miller.py
from random import randrange
import math

def mod_e(base, expo, n):
    val1 = 1
    while expo > 0:
        # the case that expo is odd, make sure to multiply by 1 base
        if expo % 2 == 1:
            val1 = (val1 * base) % n
        # repeated squaring
        base = (base * base) % n
        expo = math.floor(expo / 2)
    return val1 % n


def probably_prime(n, k):
    """
    Return True if n passes k rounds of the Miller-Rabin primality
    test (and is probably prime). Return False if n is proved to be
    composite.
    """
    small_primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31]
    if n < 2:
        return False
    for p in small_primes:
        if n < p * p:
            return True
        if n % p == 0:
            return False
    r, s = 0, n - 1
    while s % 2 == 0:
        r += 1
        s //= 2
        print s
    print r
    for _ in range(k):
        a = randrange(2, n - 1)
        x = mod_e(a, s, n)
        if x == 1 or x == n - 1:
            continue
        for _ in range(r - 1):
            x = mod_e(x, 2, n)
            if x == n - 1:
                break
        else:
            return False
    return True


print probably_prime(967, 30)
